### PR TITLE
[PyTorch] Specialize Vectorized<Half> for NEON even if FP16 arithmetic isn't available

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139220
* #139208
* #139159
* #139090
* #139084
* #139083
* #139082
* #139081
* #138275
* #138005
* #137918
* #137917
* #137916
* #137915
* #137914
* #137913
* #137912
* __->__ #137911
* #137661

We can do most of what this header does (by line count) anyway by converting to and from float.

Differential Revision: [D64265757](https://our.internmc.facebook.com/intern/diff/D64265757/)

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by swolchok** [comment]
(it's hard to succinctly explain what the benefits of this one are, so I gave it a not-user-facing for release notes)

**Review by jgong5 — APPROVED** [review]

**Review by malfet — APPROVED** [review]
Do you have any before-after perf numbers?

File: aten/src/ATen/cpu/vec/vec128/vec128_half_neon.h [hunk 1]
```diff
@@ -655,24 +768,72 @@ inline void convert(const int16_t* src, float16_t* dst, int64_t n) {
...
+  const auto [c_float_low, c_float_high] = convert_half_float(c);
+  return convert_float_half(
+      fmsub(a_float_low, b_float_low, c_float_low),
+      fmsub(a_float_high, b_float_high, c_float_high));
+#endif
 }
 
-#endif /* defined(aarch64) && defined(__ARM_FEATURE_FP16_VECTOR_ARITHMETIC) && !defined(C10_MOBILE) */
+// Returns (acc_low + a_low_half * b_low_half, acc_high + a_high_half * b_high_half)
+std::pair<Vectorized<float>, Vectorized<float>> inline fmadd(
```
Comment (swolchok) [inline aten/src/ATen/cpu/vec/vec128/vec128_half_neon.h:808]: this new fmadd variant is too hardware- and data-type-specific; it's not even reasonable to implement on the same ISA for bfloat16. Removing it.

**Comment by swolchok** [comment]
> 'test/inductor/test_torchinductor_dynamic_shapes.py::DynamicShapesCpuTests::test_multi_threading_dynamic_shapes_cpu'

this failure doesn't seem to repro locally on a linux machine, nor does it pass the sniff test because this diff only affects ARM and it's on Windows...

**Pull request closed without merging**